File: Greedy/maxSubArray.cpp
/*
58. Maximum Subarray
Given an integer array nums, find the subarray
 with the largest sum, and return its sum.
*/
#include <algorithm>
using namespace std;

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int maxSubArr = nums[0];

        int currentSum = 0;

        // starting going through nums
        for(int i = 0; i < size(nums); i++){
            // if we start getting a negative subarray, reset-forget and move on to a new subarray
            if(currentSum < 0){
                currentSum = 0;
            }
            // start adding the subArray
            currentSum += nums[i];

            maxSubArr = max(maxSubArr, currentSum);
        }
        return maxSubArr;
    }
};

File: 1D-DynamicProg/maxSubarray.py
"""
152. Maximum Product Subarray

Given an integer array nums, find a subarray
that has the largest product, and return the product.

"""

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        output = max(nums) # set output to the max value in nums
        currentMin, currentMax = 1, 1

        for n in nums:
            if n == 0:
                # 0s will mess up min and max
                currentMin, currentMax = 1, 1
                continue

            temp = currentMax * n # save old value
            currentMax = max(n * currentMax, n * currentMin, n)
            currentMin = min(temp, n * currentMin, n)

            output = max(output, currentMax, currentMin)

        return output
